File: frontend/src/types/agent.ts
export interface Agent {
  id: string;
  name: string;
  description: string;
  model: string;
  systemPrompt: string;
  tools: AgentTool[];
  userId: string;
  metadata: Record<string, any>;
  createdAt: string;
  updatedAt: string;
}

export interface AgentTool {
  name: string;
  description: string;
  parameters: Record<string, any>; // JSON Schema
  required: boolean;
}

export interface AgentState {
  agents: Agent[];
  selectedAgent: Agent | null;
  loading: boolean;
  generating: boolean;
  error: string | null;
}

export interface AgentFormValues {
  name: string;
  description: string;
  model: string;
  systemPrompt: string;
  tools: AgentTool[];
  metadata: Record<string, any>;
}

export interface AgentQuery {
  agentId: string;
  query: string;
  contextId?: string;
}

export interface Message {
  id: string;
  role: 'user' | 'assistant' | 'system';
  content: string;
  timestamp: string;
}

export interface Conversation {
  id: string;
  agentId: string;
  messages: Message[];
  createdAt: string;
  updatedAt: string;
  title: string;
  userId: string;
}

export interface AgentResponse {
  message: string;
  conversationId: string;
  recommendations?: any[];  // Type depends on the specific agent's response format
  actions?: any[]; // Type depends on the specific agent's response format
  metadata?: Record<string, any>;
}

export interface AgentResponseState {
  response: AgentResponse | null;
  loading: boolean;
  error: string | null;
} 

File: backend/app/schemas/agent.py
from pydantic import BaseModel, Field
from typing import Optional, List, Dict, Any
from datetime import datetime

class Tool(BaseModel):
    name: str
    description: str
    parameters: Dict[str, Any]
    required_parameters: List[str]
    
class AgentBase(BaseModel):
    name: str
    description: str
    model: str
    system_prompt: str
    tools: Optional[List[Tool]] = []
    metadata: Optional[Dict[str, Any]] = {}
    
class AgentCreate(AgentBase):
    pass
    
class AgentUpdate(BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None
    model: Optional[str] = None
    system_prompt: Optional[str] = None
    tools: Optional[List[Tool]] = None
    metadata: Optional[Dict[str, Any]] = None
    
class AgentResponse(BaseModel):
    id: str = Field(..., alias="_id")
    name: str
    description: str
    model: str
    system_prompt: str
    tools: Optional[List[Tool]] = []
    metadata: Optional[Dict[str, Any]] = {}
    created_at: datetime
    updated_at: datetime
    
    class Config:
        orm_mode = True
        allow_population_by_field_name = True
        
class AgentQuery(BaseModel):
    query: str
    parameters: Optional[Dict[str, Any]] = {}
    user_id: Optional[str] = None
    
class AgentQueryResponse(BaseModel):
    response: str
    thinking: Optional[str] = None
    tool_calls: Optional[List[Dict[str, Any]]] = None
    metadata: Optional[Dict[str, Any]] = {} 